### ISSUE TITLE
[API Proposal]: Add API to gracefully terminate process

### ISSUE THREAD

**Issue by tmat** [dotnet/runtime#109432]
### Background and motivation

dotnet-watch needs to be able to terminate process gracefully, i.e. give it a chance to clean up resources.

On Windows this API would call `GenerateConsoleCtrlEvent`, on Unix it would send SIGTERM signal.
Currently this requires reimplementing both process creation and termination using P/Invokes.

### API Proposal

Multiple options. One option is to add parameter to `Process.Kill` that indicates whether we want forced kill or "soft" termination (Ctrl+C/SIGTERM).

```csharp
Process.Kill(bool entireProcessTree, bool forced = true);
```

ProcessStartInfo might need a bool property that allows Kill to send Ctrl+C.

### API Usage

```csharp
var process = new Process
{
    StartInfo =
    {
        // file name, arguments, output redirection, etc.
        EnableConsoleCtrlEvents = true
    }
};

process.Start();

process.Kill(entireProcessTree: false, forced: false);

process.WaitForExit();
```

**Comment by Symbai** [comment]
> `Process.Kill` that indicates whether we want forced kill or "soft" termination

The world `kill` conflicts with `soft`. Especially when doing it gracefully the other process still has the option to say "nope". Since we already have `Process.CloseMainWindow` which gracefully closes a process if it has a window, for me it makes more sense to introduce a new method, something like `Process.CloseConsole` etc. Otherwise it gets confusing especially for beginners which method they would have to use.

**Comment by karakasa** [comment]
just fyi `GenerateConsoleCtrlEvent` cannot send events to arbitrary process.

**Comment by rjgotten** [comment]
> > `Process.Kill` that indicates whether we want forced kill or "soft" termination
> 
> The word `kill` conflicts with `soft`. Especially when doing it gracefully the other process still has the option to say "nope". Since we already have `Process.CloseMainWindow` which gracefully closes a process if it has a window, for me it makes more sense to introduce a new method, something like `Process.CloseConsole` etc. Otherwise it gets confusing especially for beginners which method they would have to use.

^ This, indeed.
Instead- what seems to be needed is something like a 
```cs
public class Process
{
  async Task<bool> RequestClose(CancellationToken cancellationtoken);

  // ...
}

**Comment by fbrosseau** [comment]
> just fyi `GenerateConsoleCtrlEvent` cannot send events to arbitrary processes (unless within the same console or created with a certain flag).

The answer has more corner cases as well - if the process is attached to a modern pseudoconsole to which you have a handle to stdin, you can write 0x3 to stdin too. This is the most reliable way to implement ctrl-c today. But the underlying problem remains the same - you need specific knowledge of the child to implement this properly on Windows - there is no generic solution.

Because these scenarios require deep control of both the child's code, as well as the code actually creating the child process, usually this is best accomplished by just pinvoking CreateProcess. This opens the door to other improvements that aren't really possible to implement generically (because they're breaking changes), such as providing asynchronous pipes as stdout/stdin for both sides of the parent/child handles.

**Comment by tmat** [comment]
> The answer has more corner cases as well - if the process is attached to a modern pseudoconsole to which you have a handle to stdin, you can write 0x3 to stdin too. This is the most reliable way to implement ctrl-c today. But the underlying problem remains the same - you need specific knowledge of the child to implement this properly on Windows - there is no generic solution.

Seems like the API can support both ways of sending Ctrl+C and it can either detect what the state process is or it can take a flag so that the caller can choose. 

Reimplementing process launching via P/Invokes is not a good option. It shifts the burden of testing all the corner cases and quirks of all the supported OSes to the implementation.

**Comment by jeffhandley** [comment]
> This API proposal sounds super convenient for consumers, but as with most process-related advanced features, it's hard to implement as a framework.

I agree with that assessment, and I think this would need a substantial time investment that our team won't be able to make. I'm happy to leave this open though if others want to prototype it out. I'll mark it as https://github.com/dotnet/runtime/labels/help%20wanted.

**Comment by davidfowl** [comment]
Thus is something we hit while trying to build our support for aspire in dotnet watch in the sdk. It’s a big lift to go process.start to p-invoking. If we do have to keep pinvoke to send the ctrl+c that’s not the end of the world, we do the same on Linux for sending signals. The real problem is replicating process start to create the process within a process group on windows.


Is there some click stop where that’s what we build into the process class?

**Comment by tmat** [comment]
>  there is no generic solution.

@jeffhandley  We are not asking for a general solution. For dotnet-watch scenario we create the process and can configure it with whatever flags necessary to support Ctrl+C.

**Comment by jkotas** [comment]
Related / partial duplicate of https://github.com/dotnet/runtime/issues/44944

**Comment by jozkee** [comment]
Fixed by https://github.com/dotnet/runtime/issues/44944

**Comment by davidfowl** [comment]
👌🏾👌🏾👌🏾👌🏾👌🏾
